Nanobind

petrasvestartas opened the pull request:
This is pull request to change pybind11 build system of C++ to Python binding to nanobind.

Following methods are wrapped:

- Boundary loops
![Screenshot from 2025-03-31 16-48-14](https://github.com/user-attachments/assets/d3d64777-2325-4543-a4ac-d29bde21e32d)

- Curvature (gaussian - normals, principal, in-plane vectors)
![image](https://github.com/user-attachments/assets/67d199cc-d8ce-488b-b02d-20f10efc9196)


- Geodesic Distance 
![image](https://github.com/user-attachments/assets/81a081f5-8f55-42ec-a3ee-a38aa9f9d67e)

- Ray/Mesh Intersections 
![image](https://github.com/user-attachments/assets/80be42c3-d93b-4dab-8e29-0998986002c2)

- Isolines 
![image](https://github.com/user-attachments/assets/8983ba7e-8e17-41c8-a327-49809b6499ac)

- Mesh Mass Matrix 
![image](https://github.com/user-attachments/assets/33d7d630-c674-432a-a417-30b6271ca560)

- Mesh Parametrisations 
![image](https://github.com/user-attachments/assets/92d27b47-ab6a-4fd1-9541-6691b381e405)

- Quad Mesh Planarisation 
![image](https://github.com/user-attachments/assets/c75b261c-dfc7-4f1e-afc5-f942174b363e)

- Remeshing from isolines
![image](https://github.com/user-attachments/assets/848c1dab-39f8-4974-986e-3debe206d16d)
![image](https://github.com/user-attachments/assets/98436787-2148-4f18-ad83-0e998a442cb6)
![image](https://github.com/user-attachments/assets/dd2f1bd2-ceab-4436-aff2-8a0fb7fe8e34)

petrasvestartas commented:
First method works. 

7  more to do:

- Curvature
- Geodesic Distance
- Ray/Mesh Intersections
- Isolines
- Mesh Mass Matrix
- Mesh Parametrisations
- Quad Mesh Planarisation

tomvanmele commented:
the isolines are wrong...

petrasvestartas commented:
> the isolines are wrong...

Is it related to building process or implementation?

Isolines method has been slightly changed in the recent libigl source code. Now it takes isovalues to retrieve output.
Also a third output is added for grouping. This helps to avoid x, y, z coordinate sorting.

petrasvestartas commented:
Building libigl on all os works fine.

Windows problem was this line:
`#define NOMINMAX`

I will clean up the repository:
- documentation examples: images and renaming: rename examples to example_name.py
- add api for remeshing on isolines
- add multiple iso remeshing split example from C++
- check if cibuildwheels work correctly locally
- bring test files from the methods docstrings to tests folder

petrasvestartas commented:
@tomvanmele  please let me know if there any changes needed. 
The package build is transformed to nanobind workflow similar to compas_cgal.

Since multiple files were changed, the simplest way, would be to check the branch: https://github.com/petrasvestartas/compas_libigl/tree/nb

Locally I tested the PYPI building process on windows and docker for linux:
![Screenshot 2025-04-04 012854](https://github.com/user-attachments/assets/19604b44-a27f-4783-a614-80966e0fd2c1)

For release to pypi we need the pending publisher:
![image](https://github.com/user-attachments/assets/316b27c0-d996-4faa-af62-92a7d7cd5f4f)

tomvanmele commented:
> We see in this pull-request the whole history of commit. Should I leave it as is or should I create a new pull-request with the changes only?

perhaps a bit annoying, but i think it would be better...

tomvanmele commented on docs/examples/example_curvature_gaussian.py:
```diff
@@ -0,0 +1,70 @@
+max_gaussian = max(gaussian_curvature[i] for i in non_boundary_vertices)
+
+
+# Create two-color maps for negative and positive curvature
+cmap_negative = ColorMap.from_two_colors(Color.blue(), Color.yellow())
+cmap_positive = ColorMap.from_two_colors(Color.yellow(), Color.magenta())
+
+# Create vertex colors dictionary
+vertex_colors = {}
+for i, k in enumerate(gaussian_curvature):
+    if trimesh.is_vertex_on_boundary(i):
+        # Set boundary vertices to black
+        vertex_colors[i] = Color.black()
+    else:
+        if k < 0:
+            vertex_colors[i] = cmap_negative(k, minval=min_gaussian, maxval=0)
+        else:
+            vertex_colors[i] = cmap_positive(k, minval=0, maxval=max_gaussian)
```
not now, but we need to make things like this simpler

tomvanmele commented on docs/examples/example_curvature_gaussian.py:
```diff
@@ -0,0 +1,70 @@
+viewer.scene.add(trimesh, use_vertexcolors=True, vertexcolor=vertex_colors)
+
+# Add normal vectors scaled by Gaussian curvature
+normal_scale = -10
+for vertex in trimesh.vertices():
+    if not trimesh.is_vertex_on_boundary(vertex):
+        point = Point(*trimesh.vertex_coordinates(vertex))
+        normal = trimesh.vertex_normal(vertex)
+        k = gaussian_curvature[vertex]
+
+        scaled_normal = [n * normal_scale * k for n in normal]
+        end_point = Point(*(p + n for p, n in zip(point, scaled_normal)))
```
@Licini perhaps we can add a few special scene objects for making recurring visualisation patterns easier to execute

* mesh vertex normals
* mesh face normals
* mesh/graph paths
* ...

tomvanmele commented on docs/examples/example_intersections.rst:
```diff
@@ -1,10 +1,10 @@
 ********************************************************************************
-Ray/Mesh Intersections
+Mesh-ray intersections
```
it is the ray that intersects the mesh

tomvanmele commented on docs/examples/example_meshing_planes.py:
```diff
@@ -0,0 +1,37 @@
+from compas.colors import ColorMap
+from compas.geometry import Rotation
+from compas.geometry import Scale
+import math
```
i can't believe this passes ruff linting and formatting

tomvanmele replied:
please turn on sorting in the project config

tomvanmele commented on docs/installation.rst:
```diff
@@ -5,36 +5,29 @@ Installation
 .. code-block:: bash
 
-    conda install -c conda-forge compas_libigl
+    conda create -n libigl -c conda-forge compas_libigl
```
remove the environment part

tomvanmele replied:
add viewer here immediately and mention that it is optional

tomvanmele commented on docs/installation.rst:
```diff
@@ -5,36 +5,29 @@ Installation
 
-Get a local copy of the source code of :mod:`compas_libigl` with all submodules.
+Several examples use the COMPAS Viewer for visualisation.
+To install :mod:`compas_viewer` in the same environment
 
 .. code-block:: bash
 
-    git clone --recursive https://github.com/compas-dev/compas_libigl.git
+    conda activate libigl
+    pip install compas_viewer
```
see above

tomvanmele commented on docs/tutorial.rst:
```diff
@@ -5,112 +5,90 @@ Tutorial
+* :func:`compas_libigl.trimesh_gaussian_curvature` - Compute Gaussian curvature at vertices
+* :func:`compas_libigl.trimesh_principal_curvature` - Compute principal curvatures and directions
 
-    source = trimesh.vertex_sample(size=1)[0]
-    distance = compas_libigl.trimesh_geodistance(
-        (V, F),
-        source,
-        method="heat",
-    )
+Geodesic Distances
```
not sure i would categorize this under analysis. for me this highlights why excessive structuring is not really helpful

tomvanmele commented on src/compas_libigl/geodistance.py:
```diff
@@ -0,0 +1,81 @@
+
+    Parameters
+    ----------
+    M : tuple[:class:`list`, :class:`list`]
```
remove :class:

tomvanmele commented on src/compas_libigl/geodistance.py:
```diff
@@ -0,0 +1,81 @@
+
+    Parameters
+    ----------
+    M : tuple[:class:`list`, :class:`list`]
```
should be `tuple[list[list[float]], list[list[int]]]`

tomvanmele commented on src/compas_libigl/planarize.py:
```diff
@@ -0,0 +1,39 @@
+
+    Parameters
+    ----------
+    M : tuple[list[list[float]], list[list[int]]] | :class:`compas.datastructures.Mesh`
```
not sure if it is still true that the mesh unpacks into vertices and faces in current compas versions

petrasvestartas commented:
@tomvanmele thank you for the corrections, I fixed them and added a new pull-request to avoid commit history.

https://github.com/compas-dev/compas_libigl/pull/21

Also found the right command for ruff:
ruff check --select I --fix '/Users/petras/brg/2_code/forks/compas_libigl/docs/examples'
